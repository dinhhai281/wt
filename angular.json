{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "angular-unit-test": {
      "projectType": "application",
      "schematics": {
        "@schematics/angular:component": {
          "style": "scss",
          "changeDetection": "OnPush",
          "viewEncapsulation": "None"
        }
      },
      "root": "",
      "sourceRoot": "src",
      "prefix": "app",
      "architect": {
        "build": {
          "builder": "@angular-builders/custom-webpack:browser",
          "options": {
            "outputPath": "dist/angular-unit-test",
            "index": "src/index.html",
            "main": "src/main.ts",
            "polyfills": "src/polyfills.ts",
            "tsConfig": "tsconfig.app.json",
            "aot": true,
            "assets": [
              "src/favicon.ico",
              "src/assets"
            ],
            "styles": [
              "src/styles.scss"
            ],
            "scripts": [],
            "customWebpackConfig": {
              "path": "webpack.config.js"
            }
          },
          "configurations": {
            "production": {
              "fileReplacements": [
                {
                  "replace": "src/environments/environment.ts",
                  "with": "src/environments/environment.prod.ts"
                }
              ],
              "optimization": true,
              "outputHashing": "all",
              "sourceMap": false,
              "extractCss": true,
              "namedChunks": false,
              "extractLicenses": true,
              "vendorChunk": false,
              "buildOptimizer": true,
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "2mb",
                  "maximumError": "5mb"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "6kb",
                  "maximumError": "10kb"
                }
              ]
            }
          }
        },
        "serve": {
          "builder": "@angular-builders/custom-webpack:dev-server",
          "options": {
            "browserTarget": "angular-unit-test:build",
            "customWebpackConfig": {
              "path": "webpack.config.js"
            }
          },
          "configurations": {
            "production": {
              "browserTarget": "angular-unit-test:build:production"
            }
          }
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n",
          "options": {
            "browserTarget": "angular-unit-test:build"
          }
        },
        "lint": {
          "builder": "@angular-eslint/builder:lint",
          "options": {
            "lintFilePatterns": [
              "src/**/*.ts",
              "src/**/*.component.html"
            ],
            "exclude": [
              "**/node_modules/**"
            ]
          }
        }
      }
    }
  },
  "defaultProject": "angular-unit-test"
}
