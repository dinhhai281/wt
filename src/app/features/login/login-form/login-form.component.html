<section
  class="relative login-form bg-gray-900 text-gray-500 text-sm px-6 py-8 rounded-md grid grid-cols-8 gap-16 shadow-md"
  *ngIf="vm$ | async as vm"
>
  <form
    class="flex flex-col justify-center items-center col-span-5"
    [formGroup]="form"
    (ngSubmit)="submit$.next()"
  >
    <h1 class="text-xl font-semibold text-white mb-2">Welcome back!</h1>
    <h3 class="text-base">We're so excited to see you again!</h3>
    <div class="flex flex-col mt-4 w-full">
      <label class="mb-2 text-xs text-gray-600 font-medium" for="email">EMAIL</label>
      <input
        class="p-2 flex-grow rounded text-gray-400 focus:shadow-outline focus:outline-none bg-gray-700 border-2 border-black"
        placeholder="Enter your email"
        name="email"
        formControlName="email"
        [readOnly]="vm.loading"
      />
      <p class="ml-1 mt-1 h-4 text-red-700 text-xs">
        <span *ngIf="vm.emailError.required">Email is required</span>
        <span *ngIf="vm.emailError.email">Not a valid email</span>
      </p>

      <label class="mb-2 text-xs mt-3 text-gray-600 font-medium" for="password">PASSWORD</label>
      <input
        class="p-2 flex-grow rounded text-gray-400 focus:shadow-outline focus:outline-none bg-gray-700 border-2 border-black"
        placeholder="Enter your password"
        name="password"
        type="password"
        formControlName="password"
        [readOnly]="vm.loading"
      />
      <a class="text-indigo-400 mt-1 ml-1 text-xs cursor-pointer">Forgot your password?</a>

      <button
        class="h-10 mt-6 rounded font-medium focus:outline-none transition duration-100 text-base"
        type="submit"
        [ngClass]="vm.buttonColorClass"
        [disabled]="vm.loading"
      >
        Login
      </button>
    </div>
  </form>
  <div class="flex flex-col col-span-3 justify-center items-center">
    <div class="h-40 w-40 bg-gray-400 rounded border-4 border-white"></div>
    <h4 class="text-xl mt-4 text-gray-200">Log in with QR Code</h4>
    <p class="text-center mt-1 text-base text-gray-600">Scan this with the mobile app to login instantly.</p>
  </div>
</section>
