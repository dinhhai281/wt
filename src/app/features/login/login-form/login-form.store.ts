import { Injectable } from '@angular/core';
import { ComponentStore } from '@ngrx/component-store';
import { identity } from 'ramda';

export interface LoginFormState {
  loading: boolean;
}

@Injectable()
export class LoginFormStore extends ComponentStore<LoginFormState> {
  constructor() {
    super({ loading: false });
  }

  readonly loading$ = this.select(state => state.loading);
  readonly buttonColorClass$ = this.select(this.loading$, isLoading => ({
    'bg-indigo-400 hover:bg-indigo-500 text-gray-200': !isLoading,
    'bg-gray-300 text-gray-600': isLoading,
  }));

  readonly setLoading = this.updater<boolean>((_, isLoading) => ({
    loading: isLoading,
  }));

  readonly vm$ = this.select(
    this.loading$,
    this.buttonColorClass$,
    (loading, buttonColorClass) => ({
      loading,
      buttonColorClass,
    }),
  );
}
