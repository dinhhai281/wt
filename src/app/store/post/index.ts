export * from './post.action';
export * from './post.reducer';
export * from './post.effect';
export * from './post.selector';
export * from './post.module';
