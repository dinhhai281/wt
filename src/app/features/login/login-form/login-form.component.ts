import {
  ChangeDetectionStrategy,
  Component,
  Input,
  OnInit,
  Output,
  ViewEncapsulation,
} from '@angular/core';
import { Validators } from '@angular/forms';
import { FormBuilder, FormControl, NgValidatorsErrors } from '@ngneat/reactive-forms';
import { combineLatest, Observable, Subject } from 'rxjs';
import { filter, map, startWith, tap, withLatestFrom } from 'rxjs/operators';

import { LoginFormStore } from './login-form.store';

export interface LoginFormValue {
  email: string;
  password: string;
}

type NgError = Observable<Partial<NgValidatorsErrors>>;

@Component({
  selector: 'app-login-form',
  templateUrl: './login-form.component.html',
  styleUrls: ['./login-form.component.scss'],
  encapsulation: ViewEncapsulation.None,
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [LoginFormStore],
})
export class LoginFormComponent implements OnInit {
  @Input()
  set loading(value: boolean) {
    this.state.setLoading(value);
  }

  form = this.fb.group<LoginFormValue, NgValidatorsErrors>({
    email: this.fb.control('', [Validators.required, Validators.email]),
    password: this.fb.control(''),
  });

  vm$ = combineLatest([this.getControlErrors('email'), this.state.vm$]).pipe(
    map(([emailError, { loading, buttonColorClass }]) => ({
      emailError,
      loading,
      buttonColorClass,
    })),
  );

  submit$ = new Subject<void>();
  @Output() submitForm = this.submit$.pipe(
    tap(this.form.markAllAsTouched.bind(this.form)),
    withLatestFrom(this.form.value$, this.form.status$),
    filter(([, , status]) => status !== 'INVALID'),
    map(([, value]) => value),
  );

  getControlErrors(controlName: keyof LoginFormValue): NgError {
    const control = this.form.getControl(controlName) as FormControl<unknown>;
    return combineLatest([control.errors$ as NgError, control.touch$]).pipe(
      map(([errors, isTouch]) => (isTouch ? errors ?? {} : {})),
      startWith({}),
    );
  }

  ngOnInit() {}

  constructor(private readonly fb: FormBuilder, private readonly state: LoginFormStore) {}
}
