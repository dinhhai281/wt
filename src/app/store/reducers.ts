import * as fromPost from '@store/post';
import * as fromAuth from '@store/auth';

export interface State {
  [fromPost.POST_FEATURE_KEY]: fromPost.PostEntityState;
  [fromAuth.AUTH_FEATURE_KEY]: fromAuth.AuthState;
}

export { fromPost, fromAuth };
