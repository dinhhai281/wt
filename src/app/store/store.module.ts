import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { environment } from '@environments/environment';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { AuthFeatureStoreModule } from './application/auth';
import { PostFeatureStoreModule } from './entities/post';

@NgModule({
  declarations: [],
  imports: [
    CommonModule,
    StoreModule.forRoot(
      {},
      {
        runtimeChecks: {
          strictActionImmutability: true,
          strictStateSerializability: true,
          strictActionSerializability: true,
          strictStateImmutability: true,
          strictActionTypeUniqueness: true,
        },
      },
    ),
    EffectsModule.forRoot(),
    environment.storeInstrument,
    PostFeatureStoreModule,
    AuthFeatureStoreModule,
  ],
})
export class AppStoreModule {}
