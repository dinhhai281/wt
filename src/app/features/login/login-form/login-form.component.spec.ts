import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { render, RenderResult } from '@testing-library/angular';

import { LoginFormComponent } from './login-form.component';
import { LoginFormStore } from './login-form.store';

describe('LoginFormComponent', () => {
  let component: RenderResult<LoginFormComponent>;

  beforeEach(async () => {
    component = await render(LoginFormComponent, {
      imports: [ReactiveFormsModule],
      providers: [LoginFormStore],
    });
  });

  it('should match snapshot', () => {
    expect(component.container).toMatchSnapshot();
  });
});
