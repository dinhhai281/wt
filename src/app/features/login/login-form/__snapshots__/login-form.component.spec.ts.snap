// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginFormComponent should match snapshot 1`] = `
<div
  id="root0"
>
  
  <section
    class="relative login-form bg-gray-900 text-gray-500 text-sm px-6 py-8 rounded-md grid grid-cols-8 gap-16 shadow-md"
  >
    <form
      class="flex flex-col justify-center items-center col-span-5 ng-untouched ng-pristine ng-valid"
      novalidate=""
    >
      <h1
        class="text-xl font-semibold text-white mb-2"
      >
        Welcome back!
      </h1>
      <h3
        class="text-base"
      >
        We're so excited to see you again!
      </h3>
      <div
        class="flex flex-col mt-4 w-full"
      >
        <label
          class="mb-2 text-xs text-gray-600 font-medium"
          for="email"
        >
          EMAIL
        </label>
        <input
          class="p-2 flex-grow rounded text-gray-400 focus:shadow-outline focus:outline-none bg-gray-700 border-2 border-black ng-untouched ng-pristine ng-valid"
          formcontrolname="email"
          name="email"
          placeholder="Enter your email"
        />
        <p
          class="ml-1 mt-1 h-4 text-red-700 text-xs"
        >
          
          
        </p>
        <label
          class="mb-2 text-xs mt-3 text-gray-600 font-medium"
          for="password"
        >
          PASSWORD
        </label>
        <input
          class="p-2 flex-grow rounded text-gray-400 focus:shadow-outline focus:outline-none bg-gray-700 border-2 border-black ng-untouched ng-pristine ng-valid"
          formcontrolname="password"
          name="password"
          placeholder="Enter your password"
          type="password"
        />
        <a
          class="text-indigo-400 mt-1 ml-1 text-xs cursor-pointer"
        >
          Forgot your password?
        </a>
        <button
          class="h-10 mt-6 rounded font-medium focus:outline-none transition duration-100 text-base bg-indigo-400 hover:bg-indigo-500 text-gray-200"
          type="submit"
        >
           Login 
        </button>
      </div>
    </form>
    <div
      class="flex flex-col col-span-3 justify-center items-center"
    >
      <div
        class="h-40 w-40 bg-gray-400 rounded border-4 border-white"
      />
      <h4
        class="text-xl mt-4 text-gray-200"
      >
        Log in with QR Code
      </h4>
      <p
        class="text-center mt-1 text-base text-gray-600"
      >
        Scan this with the mobile app to login instantly.
      </p>
    </div>
  </section>
</div>
`;
